packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -3504,6 +3084,282 @@ merge_matrices(CrsMatrixStruct<Scalar, LocalOrdinal, GlobalOrdinal, Node>& Aview
   }
 }//end merge_matrices
 
+namespace ColMapFunctors
+{
+  template<typename GO, typename GView, typename bitset_t>
+  struct GatherPresentEntries
+  {
+    GatherPresentEntries(GO minGID_, const GView& gids_, bitset_t& present_)
+      : minGID(minGID_), gids(gids_), present(present_)
+    {}
+
+    KOKKOS_INLINE_FUNCTION void operator()(const GO i) const
+    {
+      present.set(gids(i) - minGID);
+    }
+
+    GO minGID;
+    GView gids;
+    bitset_t present;
+  };
+
+  template<typename GO, typename GView, typename const_bitset_t>
+  struct ListGIDs
+  {
+    ListGIDs(GO minGID_, GView& gidList_, const_bitset_t& present_)
+      : minGID(minGID_), gidList(gidList_), present(present_)
+    {}
+
+    KOKKOS_INLINE_FUNCTION void operator()(const GO i, GO& lcount, const bool& finalPass) const
+    {
+      if(finalPass && present.test(i))
+      {
+        gidList(lcount) = minGID + i;
+      }
+      if(present.test(i))
+      {
+        lcount++;
+      }
+    }
+
+    GO minGID;
+    GView gidList;
+    const_bitset_t present;
+  };
+
+  template<typename GO, typename GView, typename LView, typename LocalMapType>
+  struct ConvertGlobalToLocal
+  {
+    ConvertGlobalToLocal(const GView& gids_, LView& lids_, LocalMapType& localMap_)
+      : gids(gids_), lids(lids_), localMap(localMap_)
+    {}
+
+    KOKKOS_INLINE_FUNCTION void operator()(const GO i) const
+    {
+      lids(i) = localMap.getLocalElement(gids(i));
+    }
+
+    GView gids;
+    LView lids;
+    LocalMapType localMap;
+  };
+} //end ColMapFunctors
+
+//Build the minimal (sorted) column map for the given set of global columns
+//Then convert gids and store them in lids (gids is not modified)
+template<typename Scalar, typename LocalOrdinal, typename GlobalOrdinal, typename Node>
+Teuchos::RCP<Map<LocalOrdinal, GlobalOrdinal, Node> > AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+makeColMapAndConvertGids(GlobalOrdinal indexBase, GlobalOrdinal minCol, GlobalOrdinal maxCol,
+                   const typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::global_col_inds_array& gids,
+                   typename MMdetails::AddKernels<Scalar, LocalOrdinal, GlobalOrdinal, Node>::col_inds_array& lids,
+                   const Teuchos::RCP<const Teuchos::Comm<int>>& comm)
+{
+  using Teuchos::RCP;
+  using Teuchos::rcp;
+  int myRank = comm->getRank();
+  std::cout << "Rank " << myRank << ": column map for some subset of global cols: " << minCol << " <= col <= " << maxCol << '\n';
```

Comment:
This `std::cout` will print on all processes in the MPI communicator.  We'll need to take this out or protect it behind `Tpetra::Details::Behavior::verbose()`.  There are other unprotected `std::cout` uses in this PR.